expression/builtin_json.go

```diff
@@ -40,8 +42,120 @@ var jsonFunctionNameToPB = map[string]tipb.ExprType{
+	_ functionClass = &jsonSetFunctionClass{}
+	_ functionClass = &jsonInsertFunctionClass{}
+	_ functionClass = &jsonReplaceFunctionClass{}
+	_ functionClass = &jsonMergeFunctionClass{}
 )
 
+// argsAnyNull returns true if args contains any null.
+func argsAnyNull(args []types.Datum) bool {
+	for _, arg := range args {
+		if arg.Kind() == types.KindNull {
+			return true
+		}
+	}
+	return false
+}
+
+// datum2JSON gets or converts to JSON from datum.
+func datum2JSON(d types.Datum, sc *variable.StatementContext) (j json.JSON, err error) {
+	tp := types.NewFieldType(mysql.TypeJSON)
+	if d, err = d.ConvertTo(sc, tp); err == nil {
+		j = d.GetMysqlJSON()
+	}
+	return
+}
+
+// parsePathExprs parse strings in datums into json.PathExpression.
+func parsePathExprs(datums []types.Datum) ([]json.PathExpression, error) {
+	pathExprs := make([]json.PathExpression, 0, len(datums))
+	for _, datum := range datums {
+		pathExpr, err := json.ParseJSONPathExpr(datum.GetString())
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		pathExprs = append(pathExprs, pathExpr)
+	}
+	return pathExprs, nil
+}
+
+// createJSONFromDatums creates JSONs from Datums.
+func createJSONFromDatums(datums []types.Datum) ([]json.JSON, error) {
+	jsons := make([]json.JSON, 0, len(datums))
+	for _, datum := range datums {
+		// TODO Here sematic of creating JSON from datum is same with types.compareMysqlJSON.
+		// But it's different from "CAST sematic" because for string, cast will parse it into
+		// JSON but here and compareMysqlJSON needs a JSON with the string as primitives.
+		// We should rewrite this as a function for here and compareMysqlJSON both can use it.
+		var j json.JSON
+		switch datum.Kind() {
+		case types.KindNull:
+			j = json.CreateJSON(nil)
+		case types.KindMysqlJSON:
+			j = datum.GetMysqlJSON()
+		case types.KindInt64, types.KindUint64:
+			j = json.CreateJSON(datum.GetInt64())
+		case types.KindFloat32, types.KindFloat64:
+			j = json.CreateJSON(datum.GetFloat64())
+		case types.KindMysqlDecimal:
+			f64, _ := datum.GetMysqlDecimal().ToFloat64()
+			j = json.CreateJSON(f64)
+		case types.KindString, types.KindBytes:
+			j = json.CreateJSON(datum.GetString())
+		default:
+			s, err := datum.ToString()
+			if err != nil {
+				return jsons, errors.Trace(err)
+			}
+			j = json.CreateJSON(s)
+		}
+		jsons = append(jsons, j)
+	}
+	return jsons, nil
+}
+
+// jsonModify is the portal for modify JSON with path expressions and values.
+func jsonModify(b baseBuiltinFunc, row []types.Datum, mt json.ModifyType) (j json.JSON, err error) {
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return j, errors.Trace(err)
+	}
+	if argsAnyNull(args) {
+		return j, errors.Trace(err)
+	}
+	j, err = datum2JSON(args[0], b.ctx.GetSessionVars().StmtCtx)
+	if err != nil {
+		return j, errors.Trace(err)
+	}
+
+	pes := make([]types.Datum, 0, (len(args)-1)/2)
+	vs := make([]types.Datum, 0, (len(args)-1)/2)
+	for i := 1; i < len(args); i++ {
```

Comment:
in `func (j JSON) Set(...)`, if `len(pathExprs)` != `len(values)`, error will be returned.